gdal/port/cpl_vsil_hdfs.cpp

```diff
@@ -0,0 +1,320 @@
+ *
+ * Project:  CPL - Common Portability Library
+ * Purpose:  Implement VSI large file api for HDFS
+ * Author:   James McClain, <jmcclain@azavea.com>
+ *
+ **********************************************************************
+ * Copyright (c) 2010-2015, Even Rouault <even dot rouault at mines-paris dot org>
+ * Copyright (c) 2018, Azavea
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a
+ * copy of this software and associated documentation files (the "Software"),
+ * to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense,
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included
+ * in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL
+ * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ ****************************************************************************/
+
+//! @cond Doxygen_Suppress
+
+#include <string>
+
+#include <fcntl.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <unistd.h>
+
+#include <cstring>
+
+#include "cpl_port.h"
+#include "cpl_vsi.h"
+
+#include "cpl_conv.h"
+#include "cpl_error.h"
+#include "cpl_vsi_virtual.h"
+
+#include "hdfs.h"
+
+CPL_CVSID("$Id$")
+
+const char * VSIHDFS = "/vsihdfs/";
+
+
+/************************************************************************/
+/* ==================================================================== */
+/*                        VSIHdfsHandle                               */
+/* ==================================================================== */
+/************************************************************************/
+
+class VSIHdfsHandle final : public VSIVirtualHandle
+{
+  private:
+    CPL_DISALLOW_COPY_ASSIGN(VSIHdfsHandle)
+
+    hdfsFile poFile = nullptr;
+    hdfsFS poFilesystem = nullptr;
+    std::string oFilename;
+    bool bReadOnly;
+
+  public:
+     VSIHdfsHandle(hdfsFile poFile,
+                   hdfsFS poFilesystem,
+                   const char * pszFilename,
+                   bool bReadOnly);
+    ~VSIHdfsHandle() override;
+
+    int Seek(vsi_l_offset nOffset, int nWhence) override;
+    vsi_l_offset Tell() override;
+    size_t Read(void *pBuffer, size_t nSize, size_t nMemb) override;
+    size_t Write(const void *pBuffer, size_t nSize, size_t nMemb) override;
+    vsi_l_offset Length()
```

Comment:
as it is not a one-liner, could you move the implementation of this method outside of the class declaration ?

Reply:
Done.